jetty-server/src/main/java/org/eclipse/jetty/server/AbstractHttpInput.java

```diff
@@ -0,0 +1,769 @@
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.io.EofException;
+import org.eclipse.jetty.util.component.Destroyable;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.util.thread.AutoLock;
+
+public abstract class AbstractHttpInput extends HttpInput
+{
+    private static final Logger LOG = Log.getLogger(AbstractHttpInput.class);
+
+    private final byte[] _oneByteBuffer = new byte[1];
+
+    protected final HttpChannelState _channelState;
+    protected final ContentProducer _contentProducer;
+    protected final AutoLock _contentLock = new AutoLock();
+    protected final Condition _contentLockCondition = _contentLock.newCondition();
+
+    private Eof _eof = Eof.NOT_YET;
+    private Throwable _error;
+    private ReadListener _readListener;
+    private long _firstByteTimeStamp = Long.MIN_VALUE;
+
+    public AbstractHttpInput(HttpChannelState state)
+    {
+        _channelState = state;
+        _contentProducer = new ContentProducer(this::produceRawContent);
+    }
+
+    /* HttpInput */
+
+    @Override
+    public void recycle()
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("recycle");
+        _contentProducer.recycle();
+        _eof = Eof.NOT_YET;
+        _error = null;
+        _readListener = null;
+        _firstByteTimeStamp = Long.MIN_VALUE;
+    }
+
+    @Override
+    public Interceptor getInterceptor()
+    {
+        return _contentProducer.getInterceptor();
+    }
+
+    @Override
+    public void setInterceptor(Interceptor interceptor)
+    {
+        _contentProducer.setInterceptor(interceptor);
+    }
+
+    @Override
+    public void addInterceptor(Interceptor interceptor)
+    {
+        _contentProducer.addInterceptor(interceptor);
+    }
+
+    @Override
+    public void asyncReadProduce()
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("asyncReadProduce {}", _contentProducer);
+        _contentProducer.produceRawContent();
+    }
+
+    @Override
+    public boolean addContent(Content content)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("addContent {} {}", content, _contentProducer);
+        if (_firstByteTimeStamp == Long.MIN_VALUE)
+        {
+            _firstByteTimeStamp = System.nanoTime();
+            if (_firstByteTimeStamp == Long.MIN_VALUE)
+                _firstByteTimeStamp++;
+        }
+        _contentProducer.addContent(content);
+        if (isAsync())
+            return _channelState.onContentAdded();
+        unblock();
+        return false;
+    }
+
+    @Override
+    public boolean hasContent()
+    {
+        return _contentProducer.hasRawContent();
+    }
+
+    @Override
+    public void unblock()
+    {
+        try (AutoLock lock = _contentLock.lock())
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("signalling blocked thread to wake up");
+            _contentLockCondition.signal();
```

Comment:
I think at the very least this needs to be `signalAll`.
While having two threads read from the same input is going to be highly problematic, the use of `AbstractLockedHttpInput` does mean that two threads can call `read()` and be mutually excluded.  If more than two bytes of content then arrive, one of the reads will be woken, but the other one never will.... then the other thread will block forever!

Reply:
The `ServletInputStream` contract is unclear w.r.t this. My assumption is that this `InputStream` can be used from any thread one wants, but only one thread can use it at a time.

Please note that the pre-refactoring implementation uses `notify()` in this particular case instead of `notifyAll()` so at the very least I'd say that no regression was introduced here. This doesn't mean the existing behavior is right and we can't revisit it, but it's again an area of discussion.

Reply:
I agree totally that the contract is not clean and that having two threads calling `read()` is pretty silly thing to do.   But then users always do pretty silly things and in response we can throw ReadPending, give a rubbish response, throw some other nasty exception..... but blocking forever is a really bad way to react to bad code.

Point taken about prior `notify()`, but now we realise that we could block a bad app forever, I think going to `signalAll` with a // TODO to review is a good way to avoid losing that realisation.

Reply:
Something quick and simple we could do is to call `tryLock` instead of `lock` and throw some sort of `YouCannotUseThisApiConcurrentlyException` when the lock cannot be taken.

But again, I'd add this to the list of investigations to be done.